Improved product sync feed generation engine with enhanced performance on large stores

budzanowski opened the pull request:
Feed file rework.

This is a huge change to the codebase. Sorry for that but I don't really think that splitting it up would bring many benefits except creating confusion about how some of the elements will interact. Still, all of the elements that are were not needed have been removed, things like checking if product should be synced, feed configuration ( FB settings ) creation, fields update, UI ( potentially useful ), logging and other minor parts. They will be added in the subsequent PRs after this one will be approved and merged. It is visible in the added vs removed lines count.

One of the main concepts of this PR is total removal of the old feed structure and replacing it with clear cut classes with well-defined responsibilities:

- **FeedScheduler**: a very simple class that sets up events that will trigger the feed file generation. It interacts with the Feed Generation Job by triggering the queue start.
- **FeedFileHander**: class responsible for managing and interacting with the actual feed file. Creating, writing, deleting, renaming temporary to destination, setting up folder protection.
- **FeedProductFormatter**: utility class that takes WC_Product and turns it into an array of fields for the feed file. Responsible for formatting data, extracting necessary bits into items required by the feed fields.
- **FeedDataExporter**: a class responsible for manipulating output from **FeedProductFormatter** to technical requirements of the CSV format. This code could live inside the **FeedProductFormatter** but the feed file specification allows for other than CSV formats so if we would ever want to switch this split would become useful.
-  **FeedFileUploadHandler**: class responsible for responding to Facebook feed file requests.

The classes are connected through **GenerateProductFeed** class that orchestrates the whole feed generation process.

The one cycle of feed generation goes as follows:

- `wc_facebook_for_woocommerce_feed_schedule_action` Action Scheduler action is triggered
- action triggers `GenerateProductFeed->queue_start()`
- `GenerateProductFeed->handle_start()` is executed. This is the process initial phase where we can set up folder, empty files, also the feed file gets populated with the CSV header
- `GenerateProductFeed->handle_batch_action()` is executed. This is where processing of the items is happening. Not all of the items at once. Batch processing is slecting only a subset of items. One by one products data is getting exported to array of fields. After all of the items are porcessed they are exported into CSV compatible format and written to the temporarry feed file.
- the `handle_batch_action()` is repeated by the framework, durring different AS calls, untill all of the items have been processed.
- finally `GenerateProductFeed->handle_end()` is called. The last step is to change the product feed name from temporary to final one. 

### Testing

#### Testing feed generation
After checking out this branch the following action should be available in the Pending section of Scheduled Actions in WooCommerce Status:
![image](https://user-images.githubusercontent.com/17271089/118487939-85b28000-b71b-11eb-80e0-a8dd28927240.png)

Running this action will start the process.  Shortly after in the In-progress section one should see something like this:
![image](https://user-images.githubusercontent.com/17271089/118488474-0e312080-b71c-11eb-850c-f3117849f313.png)

This means that the code is currently at the 22 batch. If the number of products in the catalog is small it is easy to miss the ongoing feed generation. In that case the beset way to check it to go to the `wp-content/uploads/facebook_for_woocommerce` folder and look for `product_catalog.csv` or `temp_product_catalog.csv`. At the end of the process the temp version gets renamed to the `product_catalog.csv` so only this version should be visible.

The feed file will not be identical compared to the old feed file because this does not implement yet a test if the product should be synced or not. 

#### Testing the feed upload
This can be done manually by entering the following url in the browser:
~~`your.site/?wc-api=facebook_for_woocommerce_get_feed&secret=123`~~ now `your.site/?wc-api=wc_facebook_get_feed_data&secret=123` (@haszari edit)

The first request will fail because of the dummy secret `1234`. After the first try, a real secret will be generated. It can be found under the following option: `wc_facebook_feed_url_secret`. Now with the proper secret, the link should start the feed file download. In the future, the secret will be generated when the feed configuration will be uploaded to the FB server settings and it will be used to generate the full download URL. This will be something automated and merchants will not need to worry about this.

haszari commented:
@budzanowski I'm assuming this branch is an improved version of #1887. My [feedback from that PR](https://github.com/woocommerce/facebook-for-woocommerce/pull/1887#issuecomment-833229910) still applies - in particular, we should remove all UI and handle that as follow up (not necessarily a priority).

So the bigger question - of the improvements in this PR, what can we ship in 2.5.0?

In the meantime, I've retitled the PR to focus on the most important goal as I see it. Let me know if this doesn't align with your plans - we can tweak.

haszari reviewed (commented):
Thanks for your dedicated work to improve this critical component, @budzanowski ! This is going to need review and serious testing from the whole squad, as there is potentially significant impact.

Great to see the engine broken down into classes to handle each aspect - thanks for documenting how this works upfront in the PR description. 

I did a quick code review and added some comments. I'm still investigating, learning and testing to understand the feed features we're replacing, and how this will impact them (hopefully not at all - just silently improve performance). 

Here's a list of other things I noticed which might need follow up or clarification. 

- There are some POT file changes on this branch - remove.
- `@codingStandardsIgnoreLine` / `phpcs:ignore` - please add a comment explaining why it's ok to ignore, or ideally we can fix these properly.

And some other areas that I want to understand better to make sure we're ok to ship:

- Are we still planning on building on Woo CSV exporter? If not, keen to understand why and have this documented.
- The inputs and outputs (interface/UI/behaviour) of the existing feed engine, and how we are replacing them. I.e. how this will impact merchants - the nuts and bolts of what to test. From there we will need to test various upgrade scenarios - e.g. if a site is using old feed, and upgrades, confirm everything continues to work.

Also some "features" that I didn't know about - we'll need to confirm how we proceed. For example, if feature is important and used by all stores, or an optional extra.

- Password/secret protecting the feed file request. I am curious if this is essential, I'm guessing this is how it worked previously.
- Support for web hosts that don't allow `passthru`.

budzanowski commented on facebook-commerce.php:
```diff
@@ -862,7 +846,6 @@ public function load_assets() {
 			format: 'csv'
 		},
 		feedPrepared: {
-			feedUrl: '<?php echo esc_url_raw( Feed::get_feed_data_url() ); ?>',
```
This was not used for a long time as far as I was able to check. The whole `facebookAdsToolboxConfig` is a candidate for removal.

budzanowski replied:
Created an issue for this: https://github.com/woocommerce/facebook-for-woocommerce/issues/1967

budzanowski commented on facebook-commerce.php:
```diff
@@ -2201,19 +2120,8 @@ public function ajax_sync_all_fb_products() {
-
-			if ( 'feed' === $method ) {
-
-				$this->sync_facebook_products_using_feed();
```
It should be safe to remove it. I have dug through the code and from what it looks it is an abandoned part of the plugin.

budzanowski commented:
@haszari

> Are we still planning on building on Woo CSV exporter? If not, keen to understand why and have this documented.

No, we are not using Woo CSV. The reason for that is that when @danielbitzer has introduced the AS framework it was no longer logical to use the Woo CSV class. The Woo CSV consisted mainly of batching code that I wanted to use and some utility functions. Since the AS framework is our batching framework the Woo CSV is no longer necessary.

> The inputs and outputs (interface/UI/behaviour) of the existing feed engine, and how we are replacing them. I.e. how this will impact merchants - the nuts and bolts of what to test. From there we will need to test various upgrade scenarios - e.g. if a site is using old feed, and upgrades, confirm everything continues to work.

There was no interface in the old feed, except some never documented ajax endpoints that should not be used either way because FB discourages on-demand upload - we should rely on the schedule. If there will really be a use for a button then the merchants will need to use the Facebook interface that now has this build in:
![image](https://user-images.githubusercontent.com/17271089/118590907-9c041e80-b7a3-11eb-8363-83c56c1d810c.png)


>Password/secret protecting the feed file request. I am curious if this is essential, I'm guessing this is how it worked previously.

This is how it worked previously and it is essential. The `wc-api` endpoint used for getting the feed file from the site to the FB database is exposed publicly. You only need to know the site URL to be able to download the file, hence the security string.

> Support for web hosts that don't allow `passthru`.

Already in the code. If `passthru` is not enabled we are reading the content of the feed file directly. This is also how it was working before.

budzanowski commented on facebook-commerce.php:
```diff
@@ -1713,70 +1696,6 @@ public function ajax_check_feed_upload_status() {
-	 * TODO: When migrating to FBE v2.0, remove above function and rename
-	 * below function to ajax_check_feed_upload_status()
-	 **/
-	public function ajax_check_feed_upload_status_v2() {
```
This is not used anywhere. This was a part of presumably some sort of old UI fragment.
From what it looks this was acting as the current progress status for the Sync Button that now uses Batch API for synchronization purposes.

budzanowski commented on includes/Feed/FeedDataExporter.php:
```diff
@@ -0,0 +1,172 @@
+	/**
+	 * Cached attributes variants.
+	 *
+	 * @since 2.6.0
```
> Every function, hook, class, and method should have a corresponding @since version associated with it (more on that below).

From https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/php/#since-section-changelogs

budzanowski commented on includes/fbproduct.php:
```diff
@@ -54,10 +54,14 @@ class WC_Facebook_Product {
-
-			$this->id                     = $wpid;
+			if ( $wpid instanceof WC_Product ) {
+				$this->id          = $wpid->get_id();
```
This is an optimization of the flow. Most of the time when we are calling this we already have a WooCommerce product on hand so instead of passing $id we can pass the whole product that will initialize internals without additional DB call. Granted that call would be 90% cached at this stage but it is still an overhead which we try to avoid since we are doing thousands of such calls while processing the catalog.

haszari replied:
Gotcha - nice! Would be good to document that with func docblock (and variable name) while you're there.

danielbitzer replied:
Nice improvement! Perhaps we could document the params for this function now that we are accepting both IDs and product objects?

budzanowski commented on includes/Jobs/JobRegistry.php:
```diff
@@ -22,10 +26,12 @@ class JobRegistry {
-		$this->generate_product_feed_job = new GenerateProductFeed( $action_scheduler_proxy );
+		$action_scheduler_proxy          = new ActionScheduler();
+		$feed_file_handler               = new FeedFileHandler();
+		$feed_data_exporter              = new FeedDataExporter();
```
> or could they be abstracted away into GenerateProductFeed

That would be actually counter abstraction since now we are injecting dependencies to `GenerateProductFeed` which makes this a bit cleaner. What we could do is to have one additional Feed class that would assembly all of the things and would be used here instead. I don't really want to do that since it would be an abstraction for the sake of abstraction without a really clear benefit. JobRegistery is a nice place to assemble the elements required for the feed job.

budzanowski commented on includes/Jobs/GenerateProductFeed.php:
```diff
@@ -18,18 +19,84 @@ class GenerateProductFeed extends AbstractChainedJob {
+	public function handle_batch_action( int $batch_number, array $args ) {
+		$this->processed_items = array();
+		parent::handle_batch_action( $batch_number, $args );
+		$this->write_processed_items_to_feed();
```
Because the framework ( correctly - because 99% this is what someone would want ) processes the items one by one but for writing reasons we need to write in one go. This behavior was triggered by a comment we got from VIP https://github.com/woocommerce/facebook-for-woocommerce/issues/1901#issuecomment-837687668.

@danielbitzer Could we make some changes to the framework to maybe have `handle_batch_action_start` and 'handle_batch_action_end` functions that would allow making this cleaner? Or this is good enough?

haszari replied:
Aha, thanks for explaining. Since this framework is tailored to our needs, would be good to align the interface to what we need. The main thing I'm concerned with here is making it clear where the state accumulates over each item, and when it's cleared out - feels like a "global variable" pattern so just want to make it super clear.

danielbitzer replied:
Interesting! I'd be ok with this solution but I'm keen to tweak the framework if it makes is easier for us to develop with it. I [opened a PR](https://github.com/woocommerce/action-scheduler-job-framework/pull/2) to add a `::process_items()` method.

If we added this I think we could something like this:

```php

	/**
	 * Processes a batch of items.
	 *
	 * @since 1.1.0
	 *
	 * @param array $items The items of the current batch.
	 * @param array $args  The args for the job.
	 *
	 * @throws Exception On error. The failure will be logged by Action Scheduler and the job chain will stop.
	 */
	protected function process_items( array $items, array $args ) {
		$processed_items = array();

		foreach ( $items as $item ) {
			$processed_items[] = $this->process_item( $item, $args );
		}

		$this->feed_file_handler->write_to_feed_temporary_file(
			$this->feed_data_exporter->format_items_for_feed( $this->processed_items )
		);
	}




```

budzanowski replied:
:shipit:

budzanowski replied:
Implemented.

budzanowski commented on includes/Feed/FeedDataExporter.php:
```diff
@@ -0,0 +1,172 @@
+	 */
+	protected $attribute_variants = array();
+
+	const CURRENTLY_PROCESSING_COUNT = 'facebook_for_woocommerce_feed_processing_count';
```
Nope, leftover from UI cleanup.

budzanowski commented on includes/Feed/FeedFileHandler.php:
```diff
@@ -0,0 +1,199 @@
+			array(
+				'base'    => $catalog_feed_directory,
+				'file'    => '.htaccess',
+				'content' => 'deny from all',
```
Existing feature but usefull. We don't want any unwanted access to the feed file.

budzanowski commented on includes/Feed/FeedFileHandler.php:
```diff
@@ -0,0 +1,199 @@
+		$files = array(
+			array(
+				'base'    => $catalog_feed_directory,
+				'file'    => 'index.html',
```
Folder protection prevents getting the folder content information.

budzanowski commented on includes/Feed/FeedFileHandler.php:
```diff
@@ -0,0 +1,199 @@
+	 */
+	public static function get_feed_secret() {
+
+		$secret = get_option( self::OPTION_FEED_URL_SECRET, '' );
```
Exiting feature. Explained in https://github.com/woocommerce/facebook-for-woocommerce/pull/1924#issuecomment-842819822

budzanowski commented on includes/Feed/FeedFileUploadHandler.php:
```diff
@@ -0,0 +1,137 @@
+ *
+ * @since 2.6.0
+ */
+class FeedFileUploadHandler {
```
I have chosen this name to match FB nomenclature which is always referring to this operation as Upload.
The distinction made by FB is `push Upload` or `pull Upload` 🤷🏻 I can change it if this explanation is not convincing enough :)

haszari replied:
Yep I think we can make this clearer :) 

I do agree it's useful to make it clear in docs/comments what this corresponds to in the Facebook UI, but words like `upload` have a different meaning on each end.

danielbitzer replied:
I think calling this class an `UploadHandler` suggests the class performs an upload e.g. via a web request, but it's actually doing the opposite by responding to a web request. `FeedApiEndpoint`?

budzanowski replied:
I don't really understand the objection to upload.  So if it is not making/handling the upload then who does that?
The procedure is called Feed Upload and this class handless our side of that procedure.

> I think calling this class an UploadHandler suggests the class performs an upload e.g. via a web request, but it's actually doing the opposite by responding to a web request.

The opposite to upload is download. I don't really think that we are doing anything like that in the class :)

> FeedApiEndpoint

Very vague connection to what the class is actually doing.

I am taking a step back. Probably it just means that I don't understand English well enough to grasp the problem with the name I have proposed 😉 @haszari @danielbitzer please help me by providing an alternative name, OK if you want `FeedApiEndpoint` so I can proceed.

haszari replied:
`FeedApiEndpoint` is a great fit IMO - let's go with that. This class implements an API endpoint to serve up the feed file. 

"Upload" means to load a file up from a satellite system (client/browser) to a remote server ("up", in the cloud). Upload implies a client-driven transfer. In this case, it's a transfer initiated by the remote end (Facebook pull). In any case, `upload` and `download` are problematic because they imply a client OR server context, which may depend on your point of view (since Facebook is a server, and the merchant woo site is also a server).

budzanowski replied:
🤷🏻

budzanowski replied:
renamed

budzanowski commented on includes/Feed/FeedFileUploadHandler.php:
```diff
@@ -0,0 +1,137 @@
+		try {
+
+			// Bail early if the feed secret is not included or is not valid.
+			if ( $this->feed_file_handler->get_feed_secret() !== Framework\SV_WC_Helper::get_requested_value( 'secret' ) ) {
```
If it would be only used by this class it would make sense to do so, but it will be used in one additional place when we will create the feed file configuration in Facebook Catalog datasources.

budzanowski commented on includes/Jobs/GenerateProductFeed.php:
```diff
@@ -18,18 +19,84 @@ class GenerateProductFeed extends AbstractChainedJob {
+	 * @param FacebookFeedFileHandler  $feed_file_handler  Feed file creation and manipulation handler.
+	 * @param FeedDataExporter         $feed_data_exporter Handling of file data.
+	 */
+	public function __construct( ActionSchedulerInterface $action_scheduler, $feed_file_handler, $feed_data_exporter ) {
```
@danielbitzer This is a part of the AS framework. Is this something we need to look into?

budzanowski replied:
From what I see this type of type hinting is available since PHP5.0: https://mlocati.github.io/articles/php-type-hinting.html

Since this is interacting with a library that was built in a certain way I would leave it as it is.

danielbitzer replied:
I think we should be type-hinting as much as possible in any newly written code. PHP 7.0 supports most type-hints and function return types. We can possible discuss this in our meeting since it affects the code base more broadly than this PR.

haszari replied:
I agree - I don't think we need to push hard, but should be consistent within each function - all args are type hinted.

budzanowski commented on includes/Jobs/GenerateProductFeed.php:
```diff
@@ -18,18 +19,84 @@ class GenerateProductFeed extends AbstractChainedJob {
+	 *
+	 * @var array $processed_items.
+	 */
+	protected $processed_items = array();
```
Done.

haszari replied:
Aha, so this is an array of arrays, each array is a csv row.

budzanowski commented on includes/Feed/FeedFileHandler.php:
```diff
@@ -0,0 +1,199 @@
+	 *
+	 * @var string
+	 */
+	public $temp_filename = 'temp_product_catalog.csv';
```
I am adding the feed secret to the filename. This is how it was done before. I thought that we could avoid that but it will make the plugin work in Multisite configurations.

> or even use an OS temporary file (tmpfile or tmpnam).

We should not do this based on[ #1901 (comment)](https://github.com/woocommerce/facebook-for-woocommerce/issues/1901#issuecomment-837687668)

haszari replied:
Thanks for clarifying (we should get a review from @WPprodigy too!). 

I think it's worth adding a bit more docs/info about exactly how the file is generated, over multiple jobs - so the lifecycle of the temp file is super clear.

budzanowski commented on includes/Feed/FeedFileUploadHandler.php:
```diff
@@ -0,0 +1,137 @@
+	 * Helper method, do not open to public.
+	 *
+	 * @since 1.11.0
+	 * @since 2.6.0 moved.
```
Fixed.

budzanowski commented on includes/Feed/FeedDataExporter.php:
```diff
@@ -0,0 +1,172 @@
+
+/**
+ * Class FeedDataExporter
+ *
```
Done.

haszari replied:
What's the output of `FeedProductFormatter`? Can we make the docs self contained, decouple the interface of this class from the output of the other?

budzanowski commented on includes/Feed/FeedFileHandler.php:
```diff
@@ -0,0 +1,199 @@
+	 *
+	 * @var string
+	 */
+	public $filename = 'product_catalog.csv';
```
Changed to include secret in the file name - the same way as the original implementation did this.

budzanowski commented on includes/Feed/FeedDataExporter.php:
```diff
@@ -0,0 +1,172 @@
+	 * @since 2.6.0
+	 * @var array $attribute_variants Array of variants attributes.
+	 */
+	protected $attribute_variants = array();
```
Added explanation comment and `@see` tag that points to the function using this.

budzanowski commented on includes/Feed/FeedProductFormatter.php:
```diff
@@ -0,0 +1,215 @@
+ */
+class FeedProductFormatter {
+
+	const FB_ADDITIONAL_IMAGES_FOR_FEED = 5;
```
Done.

budzanowski commented on includes/Feed/FeedProductFormatter.php:
```diff
@@ -0,0 +1,215 @@
+	/**
+	 * Format additional urls into the feed requirements.
+	 *
+	 * @since 2.6.0
```
https://github.com/woocommerce/facebook-for-woocommerce/pull/1924#discussion_r634034036

budzanowski commented on includes/Feed/FeedProductFormatter.php:
```diff
@@ -0,0 +1,215 @@
+	 *
+	 * @since 2.1.0
+	 *
+	 * @param array  $product_data the product data retrieved from a Woo product passed by reference.
```
Yep, `&` removed.

budzanowski commented on includes/Feed/FeedScheduler.php:
```diff
@@ -0,0 +1,62 @@
+ */
+class FeedScheduler {
+
+	const FEED_SCHEDULE_ACTION = 'wc_facebook_for_woocommerce_feed_schedule_action';
```
Done.

budzanowski commented on includes/Feed/FeedScheduler.php:
```diff
@@ -0,0 +1,62 @@
+	 *
+	 * @var boolean
+	 */
+	protected $enable_meta_export = false;
```
Leftover from WC CSV class usage. Removed.

budzanowski commented on includes/Jobs/GenerateProductFeed.php:
```diff
@@ -18,18 +19,84 @@ class GenerateProductFeed extends AbstractChainedJob {
+	protected $feed_file_handler;
+
+	/**
+	 * Feed file creation and manipulation utility.
```
Done.

budzanowski commented on includes/Feed/FeedProductFormatter.php:
```diff
@@ -0,0 +1,215 @@
+	 * Assembles product payload in feed upload for initial sync.
+	 *
+	 * @param \WC_Facebook_Product $woo_product WooCommerce product object normalized by Facebook.
+	 * @param array                $attribute_variants Passed by reference.
```
Done.

haszari commented on bin/GenerateCategories.php:
```diff
@@ -3,8 +3,6 @@
 /**
  * The main product feed handler.
  *
- * This will eventually replace \WC_Facebook_Product_Feed as we refactor and move its functionality here.
```
FYI this is also in `includes/Products/FBCategories.php` 🗑️

budzanowski replied:
This was all over the place. I am constantly finding such entries - usually without any logical connection to the Feed.

haszari commented:
Thanks for all the explanation and all the follow up fixes @budzanowski 🙌 

### Existing features - password-protected feed file
Thanks for confirming these features are not new. Since they exist in the old code, the safest path is probably to include them, especially if the code has simply been moved/copied (I think the diff looks bigger/scarier than it really is). 

At the same time, we should take the opportunity to confirm these features are still essential and document them as appropriate. 

Specifically for the password-protection, I'm curious since product data is essentially public. I can understand some stores might want to protect their feed file, or all their products. 

Follow up here:

- Ensure the "old" code of these features is deleted, either in this PR or as follow up.
- Ensure these features are documented, in the code or user docs - i.e. take ownership of the feature going forward.

### Removing unused code
Bravo for all the dead/potentially dead code you've removed in this PR (a bold move!). I think we should follow up on your initiative here more, since we're focused on this area now. If there are other opportunities to remove dead code like #1967 let's add follow up issues and see what else we can tidy up (in 2.6).

### Testing / QA
This is the most important follow up here. This PR as it stands is not a shippable feature - though it's very close now. 

I'm thinking the best way for us to confidently ship these improvements in 2.6 is to finalise one specific flow (set of steps merchants will take) for enabling and using the feed to sync their product catalog. And then we can all test this PR and subsequent PRs against that flow, and ultimately document it for merchants to use.

As I understand it, there are two remaining essential pieces for a minimally functional sync:

1. Filtering / validation of products (so excluded products aren't synced). I.e. using validator from #1933 - a follow up PR.
2. Configuring Facebook to upload the feed. The simplest way to ship this is probably to document how to set up feed, though we could also add a link out from wp-admin, or configure it automatically in Woo (#1928).

We can also consider whether feed should be opt-in, or if we might automatically set it up (in onboarding, or for existing users). 

Once we have that flow finalised, we can organise other tasks around it – testing, QA, docs/guidance, telemetry/tracks, and removing dead code. We can also implement some unit/integration tests to build more confidence in this critical feature 🤖

haszari commented:
Tested feed generation on a live test site (Jurassic Ninja) with ~3000 products. Things are looking promising!

- Have confirmed this site hits a 30 second timeout with the existing feed generation.
- The file generated without issue, and the site happily served requests while the generation jobs were progressing in the background.
- I obtained the secret, downloaded the file and it looks like a CSV full of products. 12 MB.
- Added the file URL in `Facebook > Commerce > Catalogue > Data sources` and it uploaded and validated fine.

Here are some screenshots of these moments 🎉 

<img width="1100" alt="Screen Shot 2021-05-19 at 4 04 23 PM" src="https://user-images.githubusercontent.com/4167300/118754998-2dff4a80-b8bc-11eb-8eae-99c954c159b0.png">

<img width="2147" alt="Screen Shot 2021-05-19 at 3 56 32 PM" src="https://user-images.githubusercontent.com/4167300/118754975-22ac1f00-b8bc-11eb-9704-0edd3d452f5a.png">

I saw some possible issues - will need some more careful testing to confirm (I didn't start with a known number of products).

TLDR: we need to test with an empty FB catalogue, and a known number of products and variations.

- The file has 20k lines. I don't think this is accurate, but I used Smooth Generator, so there may be a lot more variations than I expect. According to `WooCommerce > Products`, I have 3012 products. My facebook catalogue now has 15k - 22k - 35k products (depending on which bit of UI you look at!).
- Logs could be improved. There's a `*generate_feed*` log, containing many lines, each with a number on it - looks like it's logging each id somewhere?
- Profiling logs need updating - possibly our existing `generate_feed_job` profiling is running for each batch. Maybe just need to rename it so it's clear it's a batch, and possibly include the number of products.
- There are a lot more batch jobs (Action Scheduler actions) than I expected - ~1500. Though possibly this is correct if Smooth Generator made many variations per parent product. 🤷 

<img width="1576" alt="Screen Shot 2021-05-19 at 3 51 56 PM" src="https://user-images.githubusercontent.com/4167300/118755674-69e6df80-b8bd-11eb-80f9-226bc1f69a2b.png">
<img width="1534" alt="Screen Shot 2021-05-19 at 3 52 09 PM" src="https://user-images.githubusercontent.com/4167300/118755676-6b180c80-b8bd-11eb-8134-32bce3725053.png">

<img width="1100" alt="Screen Shot 2021-05-19 at 4 04 23 PM" src="https://user-images.githubusercontent.com/4167300/118755702-7703ce80-b8bd-11eb-9539-085f4318d875.png">

<img width="632" alt="Screen Shot 2021-05-19 at 3 59 21 PM" src="https://user-images.githubusercontent.com/4167300/118755721-7e2adc80-b8bd-11eb-8449-31f9d8096378.png">
<img width="1641" alt="Screen Shot 2021-05-19 at 4 16 14 PM" src="https://user-images.githubusercontent.com/4167300/118755752-8d118f00-b8bd-11eb-953c-e90c68e4c31f.png">
<img width="1564" alt="Screen Shot 2021-05-19 at 4 16 55 PM" src="https://user-images.githubusercontent.com/4167300/118755814-b0d4d500-b8bd-11eb-85d3-05796d3b7833.png">

budzanowski commented:
@haszari
> Specifically for the password-protection, I’m curious since product data is essentially public. I can understand some stores might want to protect their feed file, or all their products.

There are a lot of things that merchants may not want to share, like hidden catalog items ( that will become available when availability will be > 0 ), promotion dates. Besides other shops could harvest descriptions and images automatically using feed content, granted that is possible now but why make it simpler for anyone.

> The file has 20k lines. I don’t think this is accurate, but I used Smooth Generator, so there may be a lot more variations than I expect.

That is my experience with Smooth Generator: a tone of variations.

> Profiling logs need updating - possibly our existing generate_feed_job profiling is running for each batch. Maybe just need to rename it so it's clear it's a batch, and possibly include the number of products.

Renamed https://github.com/woocommerce/facebook-for-woocommerce/pull/1924/commits/9d681000dcc5b5fc7a9f698ec5ddd42c177321b3 . I am not adding the number of products since this is a know parameter of the framework that we set.

> Logs could be improved. There's a *generate_feed* log, containing many lines, each with a number on it - looks like it's logging each id somewhere?

We don't need this at all. Leftover from when the framework was not doing any real processing. Removed in https://github.com/woocommerce/facebook-for-woocommerce/pull/1924/commits/0d09bdecbcffb09e8979f4ee1497aaf15ff67bc2

> There are a lot more batch jobs (Action Scheduler actions) than I expected - ~1500. Though possibly this is correct if Smooth Generator made many variations per parent product. 🤷

10 items per batch so this number makes sense. We will adjust the size of the batch when we will have the full flow ready.

> Follow up here:
Ensure the "old" code of these features is deleted, either in this PR or as follow up.
Ensure these features are documented, in the code or user docs - i.e. take ownership of the feature going forward.

Separate PR is preferable - I don't want to make the changes any bigger.

>As I understand it, there are two remaining essential pieces for a minimally functional sync:

>Filtering / validation of products (so excluded products aren't synced). I.e. using validator from #1933 - a follow up PR.
Configuring Facebook to upload the feed. The simplest way to ship this is probably to document how to set up feed, though we could also add a link out from wp-admin, or configure it automatically in Woo (#1928).

Yes :)

budzanowski commented:
> I saw some possible issues - will need some more careful testing to confirm (I didn't start with a known number of products).

We can't really test this yet since we need the validator class.

danielbitzer commented on includes/Feed/FeedProductFormatter.php:
```diff
@@ -0,0 +1,217 @@
+		$feed_row                              = array();
+		$feed_row['id']                        = $product_data['retailer_id'];
+		$feed_row['title']                     = $woo_product->woo_product->get_title();
+		$feed_row['description']               = static::get_value_from_product_data( $product_data, 'description' );
```
If there ever was a reason to use PHP short tenary this function is it 😉

```suggestion
		$feed_row['description']               = $product_data['description'] ?: '';
```

rcstr commented on facebook-commerce.php:
```diff
@@ -2389,121 +2297,6 @@ private function sync_facebook_products_using_background_processor() {
-	 *
-	 * @internal
-	 */
-	public function ajax_sync_all_fb_products_using_feed() {
```
There's a reference in `WC_Facebook_Integration_Test` to this function, worth removing even if it's a test

budzanowski replied:
Done.

rcstr commented on facebook-commerce.php:
```diff
@@ -3753,129 +3546,6 @@ function ajax_display_test_result() {
-	 *
-	 * @deprecated 1.10.0
-	 */
-	function ajax_schedule_force_resync() {
-
-		wc_deprecated_function( __METHOD__, '1.10.0' );
-		die;
-	}
```
The function stills being called in line 335
https://github.com/woocommerce/facebook-for-woocommerce/blob/84d2e84eaa4e9667c01b2b37748386d9a4d17cd2/facebook-commerce.php#L334-L336

budzanowski replied:
Deleted.

rcstr commented on facebook-commerce.php:
```diff
@@ -3753,129 +3546,6 @@ function ajax_display_test_result() {
-	 *
-	 * @param int $offset number of seconds since the beginning of the daay
-	 */
-	public function schedule_resync( $offset ) {
```
The method is being used in

https://github.com/woocommerce/facebook-for-woocommerce/blob/bf97a40040324359069f4d0190d932e495987923/includes/Lifecycle.php#L191-L193

rcstr commented on facebook-commerce.php:
```diff
@@ -3753,129 +3546,6 @@ function ajax_display_test_result() {
-	 *
-	 * @return bool
-	 */
-	public function is_resync_scheduled() {
```
Also being used in https://github.com/woocommerce/facebook-for-woocommerce/blob/bf97a40040324359069f4d0190d932e495987923/includes/Lifecycle.php#L191-L193

budzanowski replied:
Removed. We are doing things differently now so Lifecycle updates need change.

rcstr reviewed (commented):
@budzanowski - I noticed a few things while looking at the deleted stuff, I Will add more comments as I go through the other files.

Functionality-wise, sync stills running with a store of ~1k products, but so far I MySQL Server hasn't crashed (Which was happening in 2.4.0), the machine runs fine, and not errors on my logs files. 🥳 🥳 🥳  great work!

danielbitzer commented on includes/Feed/FeedProductFormatter.php:
```diff
@@ -0,0 +1,217 @@
+	 * @param array $product_image_urls Array of urls to images.
+	 * @return string Stringified list of urls.
+	 */
+	private static function format_additional_image_url( $product_image_urls ) {
```
Personally, I'd convert the private static methods in this class to non-static methods. There doesn't seem to be any reason to keep them static.

danielbitzer commented on includes/Feed/FeedProductFormatter.php:
```diff
@@ -0,0 +1,217 @@
+	 * @since 2.6.0
+	 * @param \WC_Facebook_Product $woo_product WooCommerce product object normalized by Facebook.
+	 * @param array                $attribute_variants Array of variants attributes.
+	 * @return string product feed line data
```
This function returns an array not a string.

danielbitzer commented on includes/Feed/FeedDataExporter.php:
```diff
@@ -0,0 +1,174 @@
+	 * @return array
+	 */
+	public function get_column_names() {
+		return array(
```
I'm curious about the duplication of keys and values in this array? Is there a reason they're the same, will they be different?

budzanowski replied:
Not really, this is just a leftover from the requirements of WooCommerce CSV code. We can change it now.

danielbitzer commented on includes/Feed/FeedFileUploadHandler.php:
```diff
@@ -0,0 +1,137 @@
+	 * @return void
+	 */
+	public static function init( $feed_file_handler ) {
+		if ( null === self::$instance ) {
+			self::$instance = new self( $feed_file_handler );
+		}
```
Why are we making this class a singleton? I think it would be better to instantiate this class in our existing `facebook_for_woocommerce()` plugin singleton.

budzanowski replied:
Done.

danielbitzer commented on includes/Feed/FeedProductFormatter.php:
```diff
@@ -0,0 +1,217 @@
+		// Prepare variant column for variable products.
+		$product_data['variant'] = '';
+
+		if ( $woo_product->is_type( 'variation' ) ) {
```
I'd consider moving this whole conditional block to a separate method due to it's size. It would also provide a little more clarity as to what all this code is actually doing. E.g. `::prepare_product_variation_for_feed()`.

budzanowski replied:
Refactored.

danielbitzer commented on includes/Feed/FeedProductFormatter.php:
```diff
@@ -0,0 +1,217 @@
+	const MAX_IMAGES_PER_PRODUCT = 5;
+
+	/**
+	 * Assembles product payload in feed upload for initial sync.
```
I think this description is a bit confusing. What about something simpler...

> Formats a product to be synced to Facebook

danielbitzer commented on includes/Feed/FeedProductFormatter.php:
```diff
@@ -0,0 +1,217 @@
+/**
+ * Class responsible for formatting product for feed file.
+ */
+class FeedProductFormatter {
```
Could the product format generated by this class also be used by the batch API sync process? If so, we could rename it to a more something applicable to both sync methods e.g. `ProductSync/ProductFormatter`.

budzanowski replied:
I don't really know. Maybe it could be but this should be a separate task if we would want to do this if it makes sense. I don't think that we will touch Batch API for a long time if ever.

danielbitzer reviewed (commented):
Overall this is looking really good! A huge improvement to what we had before 👏 

I'm still wrapping my head around all this code and haven't looked at every file yet. There's a lot of feedback on this PR already so I won't leave any more comments for a day to avoid overwhelming you @budzanowski.

budzanowski commented on class-wc-facebookcommerce.php:
```diff
@@ -56,9 +56,6 @@ class WC_Facebookcommerce extends Framework\SV_WC_Plugin {
 		/** @var \SkyVerge\WooCommerce\Facebook\AJAX Ajax handler instance */
 		private $ajax;
 
-		/** @var \SkyVerge\WooCommerce\Facebook\Products\Feed product feed handler */
-		private $product_feed;
```
Nope, not needed. The logic was changed a bit.

budzanowski commented:
Ready for another round of review.

rcstr reviewed (approved):
@budzanowski - Implementation looks amazing, I tested with different environments (PHP 7.0 / 7.4 / 8.0) and was able to export to FB 1-3k products without problems, nor a single error in the logs, and very fast.

The only suggestion I have is around silent errors, I think errors shouldn't be silent and instead should be logged for further investigation.

danielbitzer commented on includes/Feed/FeedScheduler.php:
```diff
@@ -0,0 +1,57 @@
+
+	const FEED_SCHEDULE_ACTION = 'facebook_for_woocommerce_start_feed_generation';
+
+	/**
+	 * Products belonging to what category should be exported.
+	 *
+	 * @var string
+	 */
+	protected $product_category_to_export = array();
```
What this? I can't see it used anywhere.

budzanowski replied:
Good catch. It was a leftover from WC CSV code.

danielbitzer commented on includes/Feed/FeedProductFormatter.php:
```diff
@@ -0,0 +1,236 @@
+	 * @param array                $attribute_variants Array of variants attributes.
+	 * @return array[string] product feed line data
+	 */
+	public function prepare_product_for_feed( $woo_product, &$attribute_variants ) {
```
```suggestion
	public function prepare_product_for_feed( $fb_product, &$attribute_variants ) {
```

It seems like this param represents a `WC_Facebook_Product` product rather than a "Woo" one.

danielbitzer commented on includes/Feed/FeedProductFormatter.php:
```diff
@@ -0,0 +1,236 @@
+			$parent_attribute_values = array(
+				'gallery_urls'       => $gallery_urls,
+				'default_variant_id' => $variation_id,
+				'item_group_id'      => \WC_Facebookcommerce_Utils::get_fb_retailer_id( $parent_product ),
```
My IDE is flagging an issue here because `get_fb_retailer_id` expects an instance of `WC_Product` but `WC_Facebook_Product` is being passed.

danielbitzer reviewed (commented):
I've almost finished looking at every file. I will continue next week. but here are my comments from today.

budzanowski commented on includes/Feed/FeedDataExporter.php:
```diff
@@ -0,0 +1,174 @@
+	 * @param array $rows Rows of product information to export.
+	 */
+	public function format_items_for_feed( $rows ) {
+		$buffer = fopen( 'php://output', 'w' ); // phpcs:ignore WordPress.WP.AlternativeFunctions.file_system_read_fopen
+		ob_start();
+
+		foreach ( $rows as $row ) {
+			$this->export_row( $row, $buffer );
+		}
+
+		$items = ob_get_clean();
+		fclose( $buffer );
+		return $items;
```
Good idea. I took a look and it is already possible in current code, see c38b95240aa272a25f371754f2cba93524b1c20a

budzanowski commented:
I think that I have not missed anything 🤯

rcstr commented on includes/Feed/FeedProductFormatter.php:
```diff
@@ -0,0 +1,236 @@
+	 * @return string Formatted price value.
+	 */
+	private function format_price_for_feed( $value, $currency ) {
+		return (string) ( round( $value / (float) 100, 2 ) ) . $currency;
```
Didn't notice anything wrong on FB but based on [API reference](https://developers.facebook.com/docs/marketing-api/catalog-batch/reference#product-item) the format should _with a space between cost and currency._. Fixed in 62d2cdca74470df02e0591d458cb55b1e3efb734

haszari commented:
> I think that I have not missed anything 🤯

Fantastic work keeping up with all the feedback on the code @budzanowski ! I think that @danielbitzer might still be reviewing.

So now to decide next steps - what to include in 2.6, should we merge this to master, how does this impact any sites using existing feed.

budzanowski commented:
> So now to decide next steps - what to include in 2.6, should we merge this to master, how does this impact any sites using existing feed.

Hopefully, nothing changes for the sites that are using the existing feed generator. I have found while testing, that I have changed the feed URL. I have brought the old name back https://github.com/woocommerce/facebook-for-woocommerce/pull/1924/commits/729fc67954b6e46cf7357463364cb0aaece79ecc 
Someone else should test a transition from old to new. From my observations, it looks like everything works, but the second pair of eyes would be valuable.

haszari commented:
> Hopefully, nothing changes for the sites that are using the existing feed generator. I have found while testing, that I have changed the feed URL. I have brought the old name back 729fc67

Awesome @budzanowski, I was wondering if the new url/API endpoint was intended. Preserving the same URL sounds like a good idea to me :)

> Someone else should test a transition from old to new. From my observations, it looks like everything works, but the second pair of eyes would be valuable.

I agree – I think @rcstr is testing this.

haszari commented:
Noting that this PR introduces the  [Action Scheduler Job Framework package](https://github.com/woocommerce/action-scheduler-job-framework). When we release this we will also need to switch to Jetpack Autoloader in same release - #1956.
